### ISSUE TITLE
performance issues: linearity reduction for R1CS

### ISSUE THREAD

**Issue by kwantam** [circify/circ#57]
I've noticed some perf issues with linearity reduction, and Sebastian reported some in their application, too. @alex-ozdemir any chance this is a known issue? I haven't dug into it, but I should be starting to in the near future.

**Comment by alex-ozdemir** [comment]
Yeah, the runtime of that pass could be improved substantially. It's the most expensive single pass for a number of workloads. Improvements are welcome! I'm also happy to share a few pointers on where to start.

**Comment by kwantam** [comment]
I'd love to hear your thoughts on where to start, when you have a moment...

**Comment by alex-ozdemir** [comment]
Roughly in order of `benefit/(uncertainty*effort)`.

1. Doing elimination using large LCs is rare yet slow and unlikely to cascade
   into further eliminations. We could
   [restrict this
   condition](https://github.com/circify/circ/blob/master/src%2Ftarget%2Fr1cs%2Fopt.rs#L105)
   to small LCs. I tried this (threshold 100) on ZoK's sha256/1024bitPadded
   and it cut total compile time by ~33% while increasing constraint count by
   ~0.5%. You should try it on Sebastian's benchmarks, but this almost
   certainly a win.

2. I think
   [this](https://github.com/circify/circ/blob/master/src%2Ftarget%2Fr1cs%2Fopt.rs#L9)
   hash-map is spurious? Can it be a vector?
   (this past Alex's fault :(, to be clear)

3. Right now, about half of sub_in's time is in GMP. A better ff impl would
   help a lot. About 25% of the GMP time is rem_floor and another about 25% is
   allocation; both of those would probably vanish.

4. Right now, we substitute eagerly, so that we can tell immediately whether a
   monomial's coefficient goes to zero. However, if we do a symbolic
   substitution---morally equivalent to allowing (coeff, LC) pairs inside
   LCs---then we decrease the amount of up-front work. There is also a nice
   heuristic (~do all coeffs sum to zero?) for detecting when an LC may be
   zero or constant, which are ultimately the important questions during the
   elimination process.

   The question is whether this kind of laziness saves anything. It could
   probably be paired with a move away from hash-map-LCs and towards array-LCs
   to substantially improve the memory access pattern of the pass, which could
   be a win, but it's hard to say for sure w/o more investigation.

A flamegraph:

![flamegraph](https://user-images.githubusercontent.com/5809398/157179088-37bca588-966e-4b9b-8185-a58201adabb9.png)

**Comment by kwantam** [comment]
Cool, this is super useful.

I suspect there's something going haywire with large R1CS instances. It may be that allocation starts eating up lots of time. I don't have any isolated examples (yet) but we're seeing execution time blow up to multiple days in some cases. I'll keep you posted on what I find.

**Comment by kwantam** [comment]
Ah, one more thing. Representing `Lc`s as HashMaps is just a convenient way to implement sparsity, right? But in most cases there's not a huge need for random access to the coeffs, which means that arrays of `(idx, coeff)` pairs would also work, no?

**Comment by alex-ozdemir** [comment]
Sounds good. One more comment about the small instance I investigated: I saw a *huge* number of substitutions of LCs with a single monomial. e.g.:`v1 -> a*v2 + b`. It might be worth digging deeper into that. There might be some improvement possible in a previous (less expensive) pass.

**Comment by alex-ozdemir** [comment]
> Ah, one more thing. Representing Lcs as HashMaps is just a convenient way to implement sparsity, right? But in most cases there's not a huge need for random access to the coeffs, which means that arrays of (idx, coeff) pairs would also work, no?

Yep! Indeed, random access is not so important, though detecting constant LCs is.
